subprojects/plugins/src/main/java/org/gradle/api/tasks/Upload.java

```diff
@@ -16,143 +16,33 @@
-import org.gradle.api.artifacts.dsl.RepositoryHandler;
-import org.gradle.api.file.FileCollection;
 import org.gradle.api.internal.ConventionTask;
-import org.gradle.api.internal.artifacts.ArtifactPublicationServices;
-import org.gradle.api.internal.artifacts.ArtifactPublisher;
-import org.gradle.api.internal.artifacts.Module;
-import org.gradle.api.internal.artifacts.configurations.ConfigurationInternal;
-import org.gradle.api.internal.artifacts.repositories.PublicationAwareRepository;
-import org.gradle.internal.Transformers;
-import org.gradle.internal.deprecation.DeprecationLogger;
-import org.gradle.util.internal.ConfigureUtil;
+import org.gradle.api.internal.DocumentationRegistry;
 import org.gradle.work.DisableCachingByDefault;
 
-import javax.annotation.Nullable;
 import javax.inject.Inject;
-import java.io.File;
-import java.util.List;
-
-import static org.gradle.util.internal.CollectionUtils.collect;
 
 /**
  * Uploads the artifacts of a {@link Configuration} to a set of repositories.
  *
  * @deprecated This class is scheduled for removal in Gradle 8.0. To upload artifacts, use the maven-publish plugin instead.
  */
-@Deprecated
+@Deprecated // TODO:Finalize Upload Removal - Issue #21439
 @DisableCachingByDefault(because = "Produces no cacheable output")
 public class Upload extends ConventionTask {
-
-    private Configuration configuration;
-    private boolean uploadDescriptor;
-    private File descriptorDestination;
-    private RepositoryHandler repositories;
+    private final DocumentationRegistry documentationRegistry;
 
     @Inject
-    protected ArtifactPublicationServices getPublicationServices() {
-        throw new UnsupportedOperationException();
+    public Upload(DocumentationRegistry documentationRegistry) {
+        this.documentationRegistry = documentationRegistry;
     }
 
     @TaskAction
     protected void upload() {
-        getLogger().info("Publishing configuration: {}", configuration);
-        DeprecationLogger.deprecateTaskType(Upload.class, getPath()).willBeRemovedInGradle8().withUpgradeGuideSection(7, "upload_task_deprecation").nagUser();
-
-        Module module = ((ConfigurationInternal) configuration).getModule();
-
-        ArtifactPublisher artifactPublisher = getPublicationServices().createArtifactPublisher();
-        File descriptorDestination = isUploadDescriptor() ? getDescriptorDestination() : null;
-        List<PublicationAwareRepository> publishRepositories = collect(getRepositories(), Transformers.cast(PublicationAwareRepository.class));
-
-        try {
-            artifactPublisher.publish(publishRepositories, module, configuration, descriptorDestination);
-        } catch (Exception e) {
-            throw new PublishException(String.format("Could not publish configuration '%s'", configuration.getName()), e);
-        }
-    }
-
-    /**
-     * Specifies whether the dependency descriptor should be uploaded.
-     */
-    @Input
-    public boolean isUploadDescriptor() {
```

Comment:
For maximum compatibility, I think we should keep all of the public methods below.

That way, if someone is doing:
```
tasks.withType(Upload) {
   uploadDescriptor = true
}
```
Their build won't blow up with a method missing exception.